### DISCUSSION TITLE
Propagate tags to SAM translated resources

### DISCUSSION THREAD

**Discussion by abbylane** [aws/serverless-application-model#2664] [Ideas]
### Describe your idea/feature/enhancement

I would like it if the SAM translator propagated tags from SAM resources to CloudFormation resources. For example, if I have a `AWS::Serverless::Function` resource in my CFT, the transformed resources such as `AWS::CodeDeploy::DeploymentGroup` do not contain these tags. 

Use Case: We would like to validate tags for SAM templates. We transform the templates before validation, and want to enforce tag validations for the transformed resources. 

### Proposal

When translating resources from SAM to CloudFormation, save the tags from the SAM resource and apply to all resulting transformed resources that support tagging.

Things to consider:
[ ] The [SAM documentation](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-specification.html) will need to be updated

**Comment by aahung** [comment]
Thanks for proposing it. There are some questions I can think of that prevent us from doing so:
- are tags meant to be propagated?
- some resources are generated by SAM from interaction between multiple resources, tags from which origin resource should be propagated?

### Are tags meant to be propagated?

Tags are metadata, and it can represent anything. If a tag is very specific and only apply to that single resource, it doesn't make sense to propagate. For example, a tag "arm-ready-lambda-function=yes" on a lambda function should not be propagated to the generated resource such as IAM role or lambda permissions.

### Multiple origin resource contributes to a single generated resource.

Imagine the template has an "AWS::Serverless::Api" and "AWS::Serverless::Function" and they both contribute to the generation of a lambda permission. There could be conflict when we propagate both resources' tags to the generated resource.

**Comment by GavinZZ** [comment]
Hi there, we've added support for `PropagateTags` property to multiple SAM resources. You should be able to find documentation on this new property on [`AWS::Serverless::Function`](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-resource-function.html#sam-function-propagatetags), [`AWS::Serverless::Api`](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-resource-api.html#sam-api-propagatetags), [`AWS::Serverless::HttpApi`](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-resource-httpapi.html#sam-httpapi-propagatetags), [`AWS::Serverless:StateMachine`](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-resource-statemachine.html#sam-statemachine-propagatetags).

When you set the property to `true`, it will automatically propagate the tags you added to the generated resources.

**Comment by GavinZZ** [comment]
Marking this discussion as resolved. Feel free to re-open if you have any additional questions.
